pkg/render/certs.go

```diff
@@ -38,37 +39,33 @@ func GetSecret(ctx context.Context, client client.Client, name string, ns string
 	return secret, nil
 }
 
-// EnsureCertificateSecret ensures that the certificate in the provided
-// secrets has the expected DNS names. If no key secret is provided, a new key
-// secret is created and returned. If the key secret provided does have the
-// right DNS name, then that given key secret is returned.
-// Otherwise a new key secret is created and returned.
-func EnsureCertificateSecret(ctx context.Context, secretName string, secret *corev1.Secret, svcDNSNames ...string) (*corev1.Secret, error) {
+// EnsureCertificateSecret ensures that the certificate in the
+// secret has the expected DNS names. If no secret is provided, a new
+// secret is created and returned. If the secret does have the
+// right DNS name then the secret is returned.
+// Otherwise a new secret is created and returned.
+func EnsureCertificateSecret(ctx context.Context, secretName string, secret *corev1.Secret, keyName string, certName string, certDuration time.Duration, svcDNSNames ...string) (*corev1.Secret, error) {
 	var err error
 
 	// Create the secret if it doesn't exist.
 	if secret == nil {
-		secret, err = render.CreateOperatorTLSSecret(nil,
-			secretName, "tls.key", "tls.crt",
-			render.DefaultCertificateDuration, nil, svcDNSNames...,
+		secret, err = CreateOperatorTLSSecret(nil,
+			secretName, keyName, certName,
+			certDuration, nil, svcDNSNames...,
 		)
 		if err != nil {
 			return nil, err
 		}
 		return secret, nil
 	}
 
-	// If the cert's DNS names have changed then we need to recreate the secret.
+	// If the cert's DNS names have changed or if the cert is invalid, create
+	// a new one.
 	ok, err := secretHasExpectedDNSNames(secret, svcDNSNames)
-
-	if err != nil {
-		return nil, err
-	}
-	// DNS names on the cert do not match expected values; create a new cert.
-	if !ok {
-		return render.CreateOperatorTLSSecret(nil,
-			secretName, "tls.key", "tls.crt",
-			render.DefaultCertificateDuration, nil, svcDNSNames...,
+	if !ok || err != nil {
```

Comment:
I'm concerned about `err != nil`, does that mean if the user provided a certificate incorrectly we would just create a new one? That isn't the right behavior, that should be an error reported through tigerastatus.

Reply:
You're right, I've fixed this I believe

Reply:
I think this is still a problem. `EnsureCertificateSecret` still will recreate the secret if the user had provided an invalid certificate, or previously provided one that is now invalid. If I'm understanding this correctly.
I think this problem is at odds with the change you are trying to make. Unless we can tell the difference between a user created Secret/Certificate and one that the operator has created I'm not quite sure how to handle the problem I'm trying to call out.

Reply:
This is addressed in later commits